File: tags/Lucene.Net_2_9_4_RC2/lib/NUnit.org/NUnit/2.5.9/samples/vb/money/MoneyBag.vb
' ****************************************************************
' This is free software licensed under the NUnit license. You
' may obtain a copy of the license as well as information regarding
' copyright ownership at http://nunit.org/?p=license&r=2.4.
' ****************************************************************

Option Explicit On 

Namespace NUnit.Samples

    Public Class MoneyBag
        Implements IMoney

        Private fmonies As ArrayList = New ArrayList(5)

        Private Sub New()

        End Sub

        Public Sub New(ByVal bag As Money())
            For Each m As Money In bag
                If Not m.IsZero Then
                    AppendMoney(m)
                End If
            Next
        End Sub

        Public Sub New(ByVal m1 As Money, ByVal m2 As Money)

            AppendMoney(m1)
            AppendMoney(m2)

        End Sub

        Public Sub New(ByVal m As Money, ByVal bag As MoneyBag)
            AppendMoney(m)
            AppendBag(bag)
        End Sub

        Public Sub New(ByVal m1 As MoneyBag, ByVal m2 As MoneyBag)
            AppendBag(m1)
            AppendBag(m2)
        End Sub

        Public Function Add(ByVal m As IMoney) As IMoney Implements IMoney.Add
            Return m.AddMoneyBag(Me)
        End Function

        Public Function AddMoney(ByVal m As Money) As IMoney Implements IMoney.AddMoney
            Return New MoneyBag(m, Me).Simplify
        End Function

        Public Function AddMoneyBag(ByVal s As MoneyBag) As IMoney Implements IMoney.AddMoneyBag
            Return New MoneyBag(s, Me).Simplify()
        End Function

        Private Sub AppendBag(ByVal aBag As MoneyBag)
            For Each m As Money In aBag.fmonies
                AppendMoney(m)
            Next
        End Sub

        Private Sub AppendMoney(ByVal aMoney As Money)

            Dim old As Money = FindMoney(aMoney.Currency)
            If old Is Nothing Then
                fmonies.Add(aMoney)
                Return
            End If
            fmonies.Remove(old)
            Dim sum As IMoney = old.Add(aMoney)
            If (sum.IsZero) Then
                Return
            End If
            fmonies.Add(sum)
        End Sub

        Private Function Contains(ByVal aMoney As Money) As Boolean
            Dim m As Money = FindMoney(aMoney.Currency)
            Return m.Amount.Equals(aMoney.Amount)
        End Function

        Public Overloads Overrides Function Equals(ByVal anObject As Object) As Boolean
            If IsZero Then
                If TypeOf anObject Is IMoney Then
                    Dim aMoney As IMoney = anObject
                    Return aMoney.IsZero
                End If
            End If

            If TypeOf anObject Is MoneyBag Then
                Dim aMoneyBag As MoneyBag = anObject
                If Not aMoneyBag.fmonies.Count.Equals(fmonies.Count) Then
                    Return False
                End If

                For Each m As Money In fmonies
                    If Not aMoneyBag.Contains(m) Then
                        Return False
                    End If

                    Return True
                Next
            End If

            Return False
        End Function

        Private Function FindMoney(ByVal currency As String) As Money
            For Each m As Money In fmonies
                If m.Currency.Equals(currency) Then
                    Return m
                End If
            Next

            Return Nothing
        End Function

        Public Overrides Function GetHashCode() As Int32
            Dim hash As Int32 = 0
            For Each m As Money In fmonies
                hash += m.GetHashCode()
            Next
            Return hash
        End Function

        Public ReadOnly Property IsZero() As Boolean Implements IMoney.IsZero
            Get
                Return fmonies.Count.Equals(0)
            End Get
        End Property

        Public Function Multiply(ByVal factor As Integer) As IMoney Implements IMoney.Multiply
            Dim result As New MoneyBag
            If Not factor.Equals(0) Then
                For Each m As Money In fmonies
                    result.AppendMoney(m.Multiply(factor))
                Next
            End If
            Return result
        End Function

        Public Function Negate() As IMoney Implements IMoney.Negate
            Dim result As New MoneyBag
            For Each m As Money In fmonies
                result.AppendMoney(m.Negate())
            Next
            Return result
        End Function

        Private Function Simplify() As IMoney
            If fmonies.Count.Equals(1) Then
                Return fmonies(0)
            End If
            Return Me
        End Function


        Public Function Subtract(ByVal m As IMoney) As IMoney Implements IMoney.Subtract
            Return Add(m.Negate())
        End Function
    End Class

End Namespace

File: lib/NUnit.org/NUnit/2.5.9/samples/csharp/money/MoneyBag.cs
// ****************************************************************
// This is free software licensed under the NUnit license. You
// may obtain a copy of the license as well as information regarding
// copyright ownership at http://nunit.org/?p=license&r=2.4.
// ****************************************************************

namespace NUnit.Samples.Money 
{

	using System;
	using System.Collections;
	using System.Text;

	/// <summary>A MoneyBag defers exchange rate conversions.</summary>
	/// <remarks>For example adding 
	/// 12 Swiss Francs to 14 US Dollars is represented as a bag 
	/// containing the two Monies 12 CHF and 14 USD. Adding another
	/// 10 Swiss francs gives a bag with 22 CHF and 14 USD. Due to 
	/// the deferred exchange rate conversion we can later value a 
	/// MoneyBag with different exchange rates.
	///
	/// A MoneyBag is represented as a list of Monies and provides 
	/// different constructors to create a MoneyBag.</remarks>
	class MoneyBag: IMoney 
	{
		private ArrayList fMonies= new ArrayList(5);

		private MoneyBag() 
		{
		}
		public MoneyBag(Money[] bag) 
		{
			for (int i= 0; i < bag.Length; i++) 
			{
				if (!bag[i].IsZero)
					AppendMoney(bag[i]);
			}
		}
		public MoneyBag(Money m1, Money m2) 
		{
			AppendMoney(m1);
			AppendMoney(m2);
		}
		public MoneyBag(Money m, MoneyBag bag) 
		{
			AppendMoney(m);
			AppendBag(bag);
		}
		public MoneyBag(MoneyBag m1, MoneyBag m2) 
		{
			AppendBag(m1);
			AppendBag(m2);
		}
		public IMoney Add(IMoney m) 
		{
			return m.AddMoneyBag(this);
		}
		public IMoney AddMoney(Money m) 
		{
			return (new MoneyBag(m, this)).Simplify();
		}
		public IMoney AddMoneyBag(MoneyBag s) 
		{
			return (new MoneyBag(s, this)).Simplify();
		}
		private void AppendBag(MoneyBag aBag) 
		{
			foreach (Money m in aBag.fMonies)
				AppendMoney(m);
		}
		private void AppendMoney(Money aMoney) 
		{
			IMoney old= FindMoney(aMoney.Currency);
			if (old == null) 
			{
				fMonies.Add(aMoney);
				return;
			}
			fMonies.Remove(old);
			IMoney sum= old.Add(aMoney);
			if (sum.IsZero) 
				return;
			fMonies.Add(sum);
		}
		private bool Contains(Money aMoney) 
		{
			Money m= FindMoney(aMoney.Currency);
			return m.Amount == aMoney.Amount;
		}
		public override bool Equals(Object anObject) 
		{
			if (IsZero)
				if (anObject is IMoney)
					return ((IMoney)anObject).IsZero;
            
			if (anObject is MoneyBag) 
			{
				MoneyBag aMoneyBag= (MoneyBag)anObject;
				if (aMoneyBag.fMonies.Count != fMonies.Count)
					return false;
                
				foreach (Money m in fMonies) 
				{
					if (!aMoneyBag.Contains(m))
						return false;
				}
				return true;
			}
			return false;
		}
		private Money FindMoney(String currency) 
		{
			foreach (Money m in fMonies) 
			{
				if (m.Currency.Equals(currency))
					return m;
			}
			return null;
		}
		public override int GetHashCode() 
		{
			int hash= 0;
			foreach (Money m in fMonies) 
			{
				hash^= m.GetHashCode();
			}
			return hash;
		}
		public bool IsZero 
		{
			get { return fMonies.Count == 0; }
		}
		public IMoney Multiply(int factor) 
		{
			MoneyBag result= new MoneyBag();
			if (factor != 0) 
			{
				foreach (Money m in fMonies) 
				{
					result.AppendMoney((Money)m.Multiply(factor));
				}
			}
			return result;
		}
		public IMoney Negate() 
		{
			MoneyBag result= new MoneyBag();
			foreach (Money m in fMonies) 
			{
				result.AppendMoney((Money)m.Negate());
			}
			return result;
		}
		private IMoney Simplify() 
		{
			if (fMonies.Count == 1)
				return (IMoney)fMonies[0];
			return this;
		}
		public IMoney Subtract(IMoney m) 
		{
			return Add(m.Negate());
		}
		public override String ToString() 
		{
			StringBuilder buffer = new StringBuilder();
			buffer.Append("{");
			foreach (Money m in fMonies)
				buffer.Append(m);
			buffer.Append("}");
			return buffer.ToString();
		}
	}
}
